pkg/sdk/validation/manifest/json_remote_implementation_test.go

```diff
@@ -0,0 +1,184 @@
+					Optional: true,
+				},
+			},
+			exprectedResults: newValidationResult(fmt.Errorf("unknown workflow input artifact \"%s\": there is no such input neither in Interface input, nor Implementation additional input", inputParametersName)),
+		},
+		"When Argo input has optional input that does exist in Interface ": {
+			ifaceInput: &hubpublicgraphql.InterfaceInput{
+				Parameters: []*hubpublicgraphql.InputParameter{
+					{
+						Name: inputParametersName,
+					},
+				},
+			},
+			implAdditionalInput: &types.AdditionalInput{},
+			argoArtifacts: []wfv1.Artifact{
+				{
+					Name:     inputParametersName,
+					Optional: true,
+				},
+			},
+			exprectedResults: newValidationResult(fmt.Errorf("invalid workflow input artifact \"%s\": it shouldn't be optional as it is defined as Interface input", inputParametersName)),
+		},
+		"When Argo input is not optional but exists in Implementation additional inputs": {
+			ifaceInput: &hubpublicgraphql.InterfaceInput{
+				Parameters: []*hubpublicgraphql.InputParameter{
+					{
+						Name: inputParametersName,
+					},
+				},
+			},
+			implAdditionalInput: &types.AdditionalInput{
+				Parameters: map[string]types.AdditionalInputParameter{
+					additionalParameterName: {},
+				},
+			},
+			argoArtifacts: []wfv1.Artifact{
+				{
+					Name:     additionalParameterName,
+					Optional: false,
+				},
+			},
+			exprectedResults: newValidationResult(fmt.Errorf("invalid workflow input artifact \"%s\": it should be optional, as it is defined as Implementation additional input", additionalParameterName)),
+		},
+		"When Argo inputs are correctly set": {
+			ifaceInput: &hubpublicgraphql.InterfaceInput{
+				Parameters: []*hubpublicgraphql.InputParameter{
+					{
+						Name: inputParametersName,
+					},
+				},
+			},
+			implAdditionalInput: &types.AdditionalInput{
+				Parameters: map[string]types.AdditionalInputParameter{
+					additionalParameterName: {},
+				},
+			},
+			argoArtifacts: []wfv1.Artifact{
+				{
+					Name:     inputParametersName,
+					Optional: false,
+				},
+				{
+					Name:     additionalParameterName,
+					Optional: true,
+				},
+			},
+			exprectedResults: newValidationResult(),
+		},
+	}
+	for tn, tc := range tests {
+		t.Run(tn, func(t *testing.T) {
+			// given
+			ctx := context.Background()
+			hubCli := fakeHubCli{
+				InterfaceRevision: &hubpublicgraphql.InterfaceRevision{
+					Spec: &hubpublicgraphql.InterfaceSpec{
+						Input: tc.ifaceInput,
+					},
+				},
+			}
+			validator := NewRemoteImplementationValidator(hubCli)
+
+			implementation := &types.Implementation{}
+			b, err := json.Marshal(tc.argoArtifacts)
+			require.NoError(t, err)
+			implementationRaw := getImplementationRaw(b)
+			require.NoError(t, json.Unmarshal(implementationRaw, implementation))
+			implementation.Spec.AdditionalInput = tc.implAdditionalInput
+
+			// when
+			result, err := validator.validateInputArtifactsNames(ctx, *implementation)
```

Comment:
Instead of definining Implementation as string (in the `getImplementationRaw`) and concatenating `argoArtifactsInput` to it, please use proper Go types. The current approach is really error-prone, and when we will use 


The current approach is complicated, and also, error-prone (as we are using string as the source for the data):
1. Provide Argo artifacts as Go Type
1. Marshal Argo artifacts to bytes slice
1. Define Implementation as string (bytes slice) and concatenate the marshalled Argo artifact bytes
1. Just to unmarshall it to `types.Implementation`

It can be simplified, and, also, we will keep static type checking

```suggestion

So it would look like this:

```suggestion
			implementation := fixImplementation(tc.ArgoArtifacts, tc.implAdditionalInput) // returns `types.Implementation`
			// when
			result, err := validator.validateInputArtifactsNames(ctx, implementation)
```

where `fixImplementation` is:

```
func getImplementationRaw(inputArtifacts []wfv1.Artifact, additionalInput *types.AdditionalInput) types.Implementation {
	return types.Implementation {
         .... // define all properties you defined in string, but using Go types instead
       }
}
```

Reply:
Here's full code: 

```go
func fixImplementation(inputArtifacts []wfv1.Artifact, implAdditionalInput *types.AdditionalInput) types.Implementation {
	workflow := argo.Workflow{
		WorkflowSpec: &wfv1.WorkflowSpec{
			Entrypoint: "test",
		},
		Templates: []*argo.Template{
			{
				Template: &wfv1.Template{
					Name: "test",
					Inputs: wfv1.Inputs{
						Artifacts: inputArtifacts,
					},
					Outputs: wfv1.Outputs{},
				},
				Steps: []argo.ParallelSteps{},
			},
		},
	}

	return types.Implementation{
		Spec: types.ImplementationSpec{
			Implements: []types.Implement{
				{
					Path:     "cap.interface.test.impl",
					Revision: "0.1.0",
				},
			},
			Action: types.Action{
				Args: map[string]interface{}{
					"workflow": workflow,
				},
			},
			AdditionalInput: implAdditionalInput,
		},
	}
}
```

And in `t.Run` you use it simply as:

```go
			implementation := fixImplementation(tc.argoArtifacts, tc.implAdditionalInput)

			// when
			result, err := validator.validateInputArtifactsNames(ctx, implementation)
```

Tested the code locally and it works 🙂

Reply:
yeah, it looks much more straightforward. Added :+1: